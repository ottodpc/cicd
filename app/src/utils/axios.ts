import axios from "axios";

const API_URL =
  process.env.REACT_APP_API_URL || "https://cfainsta.alkwelo.app/api";

const axiosInstance = axios.create({
  baseURL: API_URL,
  headers: {
    "Content-Type": "application/json",
  },
});

axiosInstance.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem("token");
    if (token) {
      config.headers["Authorization"] = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

axiosInstance.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401) {
      localStorage.removeItem("token");
      localStorage.removeItem("user");
      window.location.href = "/login";
    }
    return Promise.reject(error);
  }
);

export default axiosInstance;
