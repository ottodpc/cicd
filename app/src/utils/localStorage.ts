export const getUser = () => {
  const user = localStorage.getItem("user");
  return user ? JSON.parse(user) : null;
};

export const setUser = (user: any) => {
  localStorage.setItem("user", JSON.stringify(user));
};

export const getToken = () => {
  return localStorage.getItem("token");
};

export const setToken = (token: string) => {
  localStorage.setItem("token", token);
};

export const clearAuth = () => {
  localStorage.removeItem("user");
  localStorage.removeItem("token");
};

export const isAuthenticated = () => {
  return !!getToken();
};
