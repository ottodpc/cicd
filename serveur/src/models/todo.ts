import { Schema } from '@nestjs/mongoose'

import { Prop } from '@nestjs/mongoose'
import { ApiProperty } from '@nestjs/swagger'
import { SchemaFactory } from '@nestjs/mongoose'

@Schema()
export class Todo {
	@Prop()
	@ApiProperty()
	title: string

	@Prop()
	@ApiProperty()
	description: string

	@Prop()
	@ApiProperty()
	todoId: string

	@Prop()
	_id: string
}

export const TodoSchema: SchemaFactory = SchemaFactory.createForClass(Todo)
