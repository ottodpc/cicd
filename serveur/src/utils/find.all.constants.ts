export const regex = /(\w+):(\w+):(.*)/;

export const allowedModelSorts: Map<string, Set<string>> = new Map();

export const allowedModelFilters: Map<
  string,
  Map<string, string[]>
> = new Map();

export const fieldTypeMapper: Map<string, Map<string, string>> = new Map();

export const Operations = {
  eq: '$eq',
  gt: '$gt',
  gte: '$gte',
  lt: '$lt',
  lte: '$lte',
};

const todoModelProperty: Map<string, string> = new Map();

const allowedTodoFilters: Map<string, string[]> = new Map();

const userModelProperty: Map<string, string> = new Map();

const allowedUserFilters: Map<string, string[]> = new Map();

allowedModelSorts.set('Todo', new Set(['']));
allowedModelSorts.set('User', new Set(['']));
allowedTodoFilters.set('todoId', ['eq']);
allowedModelFilters.set('Todo', allowedTodoFilters);
allowedModelFilters.set('User', allowedUserFilters);
todoModelProperty.set('title', 'string');
todoModelProperty.set('description', 'string');
todoModelProperty.set('todoId', 'string');
fieldTypeMapper.set('Todo', todoModelProperty);
userModelProperty.set('username', 'string');
userModelProperty.set('userId', 'string');
fieldTypeMapper.set('User', userModelProperty);
