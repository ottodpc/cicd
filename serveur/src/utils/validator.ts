import { allowedModelFilters } from './find.all.constants'
import { allowedModelSorts } from './find.all.constants'

export const isFilterOperationValid = (
	filterName: string,
	operation: string,
	modelName: string
): boolean => {
	const allowedFilters = allowedModelFilters.get(modelName) || new Map()
	const allowedOperations = allowedFilters.get(filterName) || []

	if (allowedOperations.find((it) => it === operation)) {
		return true
	}
	return false
}

export const isSortValid = (modelName: string, sortField: string): boolean => {
	const allowedSorts = allowedModelSorts.get(modelName) || new Set()
	return allowedSorts.has(sortField)
}
