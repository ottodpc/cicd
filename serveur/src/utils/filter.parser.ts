import { regex } from './find.all.constants'
import { fieldTypeMapper } from './find.all.constants'
import { isFilterOperationValid } from './validator'
import { ParserException } from '../exceptions/parser.exception'
import { SearchQuery } from '../entities/searchquery'

export const filterParser = (
	filters: string,
	modelName: string
): SearchQuery[] => {
	try {
		const searchQuery: SearchQuery[] = []
		const errorMessages: string[] = []
		if (filters === '') {
			return searchQuery
		}
		filters.split(':and:').forEach((filter) => {
			const matchResult = filter.match(regex)
			const filterName = matchResult[1]
			const operation = matchResult[2]
			const filterValue = getFieldValue(
				filterName,
				matchResult[3],
				modelName
			)
			if (isFilterOperationValid(filterName, operation, modelName)) {
				searchQuery.push(
					new SearchQuery(filterName, operation, filterValue)
				)
			} else {
				errorMessages.push(
					`operation ${operation} is not allowed on ${filterName}`
				)
			}
		})

		if (errorMessages.length > 0) {
			throw new ParserException(errorMessages.join(', \n'))
		}
		return searchQuery
	} catch (e) {
		throw new ParserException('failed to parse')
	}
}

const getFieldValue = (
	fieldName: string,
	fieldValue: any,
	modelName: string
) => {
	const modelTypeMapper = fieldTypeMapper.get(modelName) || new Map()
	const type = modelTypeMapper.get(fieldName) || ''

	if (type === 'string') {
		return fieldValue.toString()
	} else if (type === 'number') {
		return parseInt(fieldValue, 10)
	} else if (type === 'Date') {
		return new Date(fieldValue)
	} else if (type === 'Boolean') {
		return fieldValue.toString().toLowerCase() == 'true'
	}
	return fieldValue.toString()
}
