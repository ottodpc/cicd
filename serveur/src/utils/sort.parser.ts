import { isSortValid } from './validator'
import { ParserException } from '../exceptions/parser.exception'
import { Sort } from '../entities/sort'

export const sortParser = (sorts: string, modelName: string): Sort[] => {
	const processedSort: Sort[] = []
	let sortDirection = ''
	let sortField = ''
	const errorMessages = []
	try {
		if (sorts === '') {
			return processedSort
		}
		sorts.split(',').forEach((entry) => {
			if (entry.charAt(0) === '+' || entry.charAt(0) === '-') {
				sortDirection = entry.charAt(0)
				sortField = entry.substring(1)
			} else {
				sortDirection = '-'
				sortField = entry
			}
			if (isSortValid(modelName, sortField)) {
				processedSort.push(new Sort(sortField, sortDirection))
			} else {
				errorMessages.push(
					`${modelName} has no sort field ${sortField}`
				)
			}
		})
	} catch (e) {
		throw new ParserException('failed to parse sort')
	}

	if (errorMessages.length > 0) {
		throw new ParserException(errorMessages.join(','))
	}

	return processedSort
}
