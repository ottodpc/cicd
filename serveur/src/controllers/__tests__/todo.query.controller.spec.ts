import { TodoService } from '../../services/todo.service';
import { TodoQueryController } from '../todo.query.controller';
import { readResourceFromExactPath } from '../../../test/utils';

describe('should test todoQueryController', () => {
  let todoQueryController: TodoQueryController;
  let todoService: TodoService;

  beforeEach(async () => {
    todoService = {
      get: jest.fn(),
      select: jest.fn(),
    } as unknown as TodoService;
    todoQueryController = new TodoQueryController(todoService);
  });

  it('should test get method', async () => {
    jest
      .spyOn(todoService, 'get')
      .mockReturnValue(
        Promise.resolve(
          await JSON.parse(
            readResourceFromExactPath(__dirname + '/resources/Todo.json'),
          ),
        ),
      );
    const actualResp = await todoQueryController.get('defaultString');
    expect(jest.spyOn(todoService, 'get')).toBeCalledTimes(1);
    expect(actualResp).toMatchSnapshot();
  });

  it('should test select method', async () => {
    jest
      .spyOn(todoService, 'select')
      .mockReturnValue(
        Promise.resolve(
          await JSON.parse(
            readResourceFromExactPath(__dirname + '/resources/Todo.json'),
          ),
        ),
      );
    const actualResp = await todoQueryController.select(
      'defaultString',
      'defaultString',
      123,
      123,
    );
    expect(jest.spyOn(todoService, 'select')).toBeCalledTimes(1);
    expect(actualResp).toMatchSnapshot();
  });
});
