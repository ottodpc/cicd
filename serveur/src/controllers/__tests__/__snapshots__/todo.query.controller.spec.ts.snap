// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test todoQueryController should test get method 1`] = `
{
  "_id": "fugiat Lorem irure",
  "description": "fugiat Lorem irure",
  "fugiat00": -16000000,
  "title": "fugiat Lorem irure",
  "todoId": "fugiat Lorem irure",
}
`;

exports[`should test todoQueryController should test select method 1`] = `
PagedResponse {
  "pageInfo": PageInfo {
    "limit": 123,
    "offset": 123,
  },
  "results": {
    "_id": "fugiat Lorem irure",
    "description": "fugiat Lorem irure",
    "fugiat00": -16000000,
    "title": "fugiat Lorem irure",
    "todoId": "fugiat Lorem irure",
  },
}
`;
